Plots/MapProjections/NCL_sat_1.py

```diff
@@ -0,0 +1,380 @@
+    # to cluster and label coordinates
+    db = DBSCAN(eps=10, min_samples=1)
+    new = db.fit(extremacoords)
+    labels = new.labels_
+
+    # Create an dictionary of values with key being coordinate
+    # and value being cluster label.
+    coordsAndLabels = {}
+    for x in range(len(extremacoords)):
+        if labels[x] in coordsAndLabels:
+            coordsAndLabels[labels[x]].append(extremacoords[x])
+        else:
+            coordsAndLabels[labels[x]] = [extremacoords[x]]
+
+    # Initialize array of coordinates to be returned
+    clusterExtremas = []
+
+    # Iterate through the coordinates in each cluster
+    for key in coordsAndLabels:
+        # Create array to hold all the field variable values for that cluster
+        datavals = []
+        for coord in coordsAndLabels[key]:
+            # Find field variable value of each coordinate
+            for x in range(len(coordarr)):
+                for y in range(len(coordarr[x])):
+                    if coordarr[x][y][0] == coord[0] and coordarr[x][y][1] == coord[1]:
+                        pval = da.data[x][y]
+            # Append the field variable value to the array for that cluster
+            datavals.append(pval)
+
+        # Find the index of the smallest/greatest field variable value of each cluster
+        if eType == 'Low':
+            index = np.argmin(np.array(datavals))
+        if eType == 'High':
+            index = np.argmax(np.array(datavals))
+
+        # Append the coordinate corresponding to that index to the array to be returned
+        clusterExtremas.append((coordsAndLabels[key][index][0], coordsAndLabels[key][index][1]))
+
+    return clusterExtremas
+
+###############################################################################
+# Helper function that will plot contour labels
+
+def plotCLabels(da, contours, transform, ax, proj, Clevels=[], lowClevels=[], highClevels=[], rfs=14, efs=22, whitebbox=False,
+                rHorizontal=False, eHorizontal=True):
+
+    """
+    Utility function to plot contour labels. Regular contour labels will be plotted using the built-in matplotlib
+    clabel function. High/Low contour labels will be plotted using text boxes for more accurate label values
+    and placement.
+    Args:
+        da: (:class:`xarray.DataArray`):
+            Xarray data array containing the lat, lon, and field variable data values.
+        contours (:class:`cartopy.mpl.contour.GeoContourSet`):
+            Contour set that is being labeled.
+        transform (:class:`cartopy._crs`):
+            Instance of CRS that represents the source coordinate system of coordinates.
+            (ex. ccrs.Geodetic()).
+        ax (:class:`matplotlib.pyplot.axis`):
+            Axis containing the contour set.
+        proj (:class:`cartopy.crs`):
+            Projection 'ax' is defined by.
+            This is the instaance of CRS that the coordinates will be transformed to.
+        Clevels (:class:`list`):
```

Comment:
It looks like rather than selecting a contour value and plotting the contour associated with that value, you are picking a location and plotting a contour for the value at that location. Is that correct? The original NCL example looks like it simply plots contours levels from 948 to 1064 in increments of 4.

Reply:
Yes- I did this because for the high and low clabels, the high and low contour values will change depending on how many levels are in the contour set. For example, if the contour set only has levels 948 to 1064 in increments of 4, if you have a high pressure value of 1075 in the data, it will probably print as 1064 because python just chooses the contour closest to the coordinate you are plotting at.

Reply:
Sorry, I didn't quite follow that explanation. But I think we're going to have to figure out how to make this work using a conventional way of specifying the contours. I.e. specifying the levels, not locations that they pass through. Perhaps you can show me an example of what happens if you try to plot the contours by level so I can better understand what the problem is?

Reply:
Ok actually I think I was misunderstanding- the contour levels are plotted by specifying levels, just like in the NCL script:

![using clevels](https://user-images.githubusercontent.com/36773444/88461533-223c2680-ce59-11ea-8789-fb90f642b527.PNG)

 I think this variable name is confusing because it actually just means a list of coordinates, it has nothing to do with the contour levels/lines that are being plotted. I will change it! Does this answer your question.

Reply:
Changing the name would help for sure. But I think the larger question is what are all of the hard-coded values in regularCLabels? Are you simply attempting to place the contour labels at the exact same locations as in the NCL plot? Is there some reason the automatic selection of labels by MPL is not adequate (other than they will lie in different locations than the original NCL plot)?

Reply:
Yes, I was just trying to place the labels in the same locations as in the NCL plot. I don't think it would make a big difference if I just used automatic MPL labeling instead, do you want me to take the manually placed labels out? And if so, do you also want me to delete the helper function that manually plots non-high/low contour labels using coordinates?

Reply:
Now I understand. In general, users are not going to want to have to manually select label locations. However, I can imagine situations when someone wants complete control over where the labels are placed. My suggestion: leave things as they are, but clearly document that this example demonstrates how to manually place contour labels. I realize the description of the plot says something like "manual placement of contour labels", but there is a lot going on in this example. I'd suggest expanding the comment for "regularCLabels" to reenforce that we're manually placing the label locations instead of letting MPL choose the locations for us. Sound good?

Reply:
Yes!